src/main/java/io/cdap/plugin/servicenow/restapi/RestAPIResponse.java

```diff
@@ -82,47 +80,47 @@ public static RestAPIResponse parse(HttpResponse httpResponse, String... headerN
         .filter(o -> headerNameList.contains(o.getName()))
         .collect(Collectors.toMap(Header::getName, Header::getValue)));
     }
-
-    String responseBody = "";
-    try {
-      responseBody = EntityUtils.toString(httpResponse.getEntity());
-    } catch (Exception e) {
-      httpStatus = HttpStatus.SC_INTERNAL_SERVER_ERROR;
-      return new RestAPIResponse(httpStatus, headers, String.format(JSON_ERROR_RESPONSE_TEMPLATE, e.getMessage()));
-    }
-
+    String responseBody = EntityUtils.toString(httpResponse.getEntity());
+    validateRestApiResponse(responseBody);
     return new RestAPIResponse(httpStatus, headers, responseBody);
   }
 
-  public static RestAPIResponse parse(HttpResponse httpResponse) {
+  public static RestAPIResponse parse(HttpResponse httpResponse) throws IOException {
     return parse(httpResponse, new String[0]);
   }
 
   public int getHttpStatus() {
     return httpStatus;
   }
 
-  public boolean isSuccess() {
-    boolean isSuccess = false;
-    // ServiceNow Rest API may return 200 OK response with an error message in the response body.
-    // Normally we would expect non-200 response code if there's an error.
-    if (this.isRetryable) {
-      isSuccess = false;
-    } else if (successCodes.contains(getHttpStatus())) {
-      isSuccess = true;
+  private static void validateRestApiResponse(String responseBody) {
+    JsonObject jo = GSON.fromJson(responseBody, JsonObject.class);
+    // check if status is "failure"
+    String status = null;
+    if (jo.get(ServiceNowConstants.STATUS) != null) {
+      status = jo.get(ServiceNowConstants.STATUS).getAsString();
+    }
+    if (!ServiceNowConstants.FAILURE.equals(status)) {
+      return;
+    }
+    // check if failure is retryable
+    String errorMessage = jo.getAsJsonObject(ServiceNowConstants.ERROR).get(ServiceNowConstants.MESSAGE).getAsString();
+    if (errorMessage.contains(ServiceNowConstants.MAXIMUM_EXECUTION_TIME_EXCEEDED)) {
+      throw new RetryableException(String.format(REST_ERROR_MESSAGE, errorMessage));
+    } else {
+      throw new NonRetryableException(String.format(REST_ERROR_MESSAGE, errorMessage));
     }
-    return isSuccess;
   }
 
-  private void checkRetryable() {
-    Gson gson = new Gson();
-    JsonObject jo = gson.fromJson(this.responseBody, JsonObject.class);
-    if (jo.get(ServiceNowConstants.STATUS) != null &&
-      jo.get(ServiceNowConstants.STATUS).getAsString().equals(ServiceNowConstants.FAILURE) &&
-      jo.getAsJsonObject(ServiceNowConstants.ERROR).get(ServiceNowConstants.MESSAGE).getAsString()
-        .contains(ServiceNowConstants.MAXIMUM_EXECUTION_TIME_EXCEEDED)) {
-      isRetryable = true;
+  private static void validateHttpResponse(HttpResponse response) {
+    int code = response.getStatusLine().getStatusCode();
+    if (SUCCESS_CODES.contains(code)) {
+      return;
+    }
+    if (RETRYABLE_CODES.contains(code)) {
```

Comment:
For retryable conditions, are we handling if access token got expired, then renew it

Reply:
For Every api call, we get a new access token. Meaning if an api call throws retryable exception, the retry call will fetch a new access token and then retry the api call.